src/main/java/xbot/common/controls/sensors/wpi_adapters/GamepadWpiAdapter.java

```diff
@@ -0,0 +1,42 @@
+package xbot.common.controls.sensors.wpi_adapters;
+
+import com.google.inject.Inject;
+import com.google.inject.assistedinject.Assisted;
+
+import edu.wpi.first.wpilibj.GenericHID;
+import edu.wpi.first.wpilibj.Joystick;
+import xbot.common.controls.sensors.XGamepad;
+import xbot.common.injection.wpi_factories.CommonLibFactory;
+import xbot.common.logging.RobotAssertionManager;
+
+public class GamepadWpiAdapter extends XGamepad {
+
+private GenericHID internalHID;
+    
+    @Inject
+    public GamepadWpiAdapter(
+            @Assisted("port") int port, 
+            @Assisted("numButtons") int numButtons,
+            CommonLibFactory clf,
+            RobotAssertionManager assertionManager) {
+        super(port, clf, assertionManager, numButtons);
+        
+        internalHID = new Joystick(port);
+    }
+    
+    protected double getX() {
```

Comment:
Why `protected`?

Reply:
See XJoystick - the primitive getX and getY are implemented by the adapters, and the abstract class uses them to produce things like vectors.